# 01 - Lectura y Discovery

**TP2 - Modulo 2 - Clasificacion `smoking`**

Primer contacto con los datos. Cargamos los dos datasets, les damos un formato comodo para trabajar y miramos que tiene adentro cada uno antes de hacer cualquier cosa.

In [1]:
import sys
from pathlib import Path

# Permite importar utils.py este en notebooks/ aunque el cwd cambie.
sys.path.insert(0, str(Path.cwd()))

import pandas as pd
import numpy as np

import utils
from utils import TRAIN_XLSX, PRED_XLSX, DATA_PROCESSED, TARGET, ID_COL

pd.set_option('display.max_columns', None)
print('ROOT del proyecto:', utils.ROOT)

ROOT del proyecto: /home/mvieira/Documentos/diplomatura/entregas/tp2


## Carga y limpieza inicial

Usamos `utils.cargar_y_sanear` que hace las cosas basicas: leer el Excel, eliminar filas vacias que vienen al final del archivo, renombrar columnas a snake_case y convertir el target a entero. La misma funcion se usa en el notebook 06 para que los datos de prediccion pasen por exactamente el mismo proceso.

In [2]:
df_train = utils.cargar_y_sanear(TRAIN_XLSX)
df_pred = utils.cargar_y_sanear(PRED_XLSX)

print('train:', df_train.shape)   # esperado ~ (50000, 27)
print('pred :', df_pred.shape)     # esperado ~ (5692, 26)

train: (50000, 35)
pred : (5692, 34)


In [3]:
df_train.head()

,id,gender,age,height_cm,weight_kg,waist_cm,eyesight_left,eyesight_right,hearing_left,hearing_right,systolic,relaxation,fasting_blood_sugar,cholesterol,triglyceride,hdl,ldl,hemoglobin,urine_protein,serum_creatinine,ast,alt,gtp,oral,dental_caries,tartar,smoking,bmi,waist_to_height,pulse_pressure,map,ast_alt_ratio,non_hdl,tg_hdl_ratio,liver_enzymes_sum
0,0,F,40,155,60,3.377083,0.043056,0.041667,0.041667,0.041667,4.750000,3.041667,3.916667,8.958333,3.416667,3.041667,5.250000,0.506250,0.041667,0.004861,0.750000,0.791667,1.125000,Y,0,Y,0,24.973975,0.021788,1.708333,3.611111,0.947367,5.916667,1.123287,2.666667
1,1,F,40,160,60,3.375000,0.005556,0.004167,0.041667,0.041667,4.958333,2.916667,5.416667,8.000000,4.791667,1.750000,5.291667,0.504861,0.041667,0.004167,0.916667,0.791667,0.750000,Y,0,Y,0,23.437491,0.021094,2.041667,3.597222,1.157893,6.250000,2.738094,2.458333
2,2,M,55,170,60,3.333333,0.005556,0.005556,0.041667,0.041667,5.750000,3.583333,3.708333,10.083333,7.583333,2.291667,6.291667,0.630556,0.041667,0.041667,0.875000,0.666667,0.916667,Y,0,N,1,20.761238,0.019608,2.166667,4.305556,1.312498,7.791667,3.309089,2.458333
3,3,M,40,165,70,3.666667,0.045139,0.045139,0.041667,0.041667,4.166667,2.500000,4.000000,13.416667,10.583333,1.875000,9.416667,0.588194,0.041667,0.041667,0.791667,1.083333,0.750000,Y,0,Y,0,25.711653,0.022222,1.666667,3.055556,0.730769,11.541667,5.644441,2.625000
4,4,F,40,155,60,3.583333,0.041667,0.041667,0.041667,0.041667,5.000000,3.083333,3.333333,7.666667,3.083333,2.583333,4.458333,0.503472,0.041667,0.004167,0.666667,0.583333,0.916667,Y,0,N,0,24.973975,0.023118,1.916667,3.722222,1.142855,5.083333,1.193548,2.166667


## Tipos de datos y memoria

In [4]:
df_train.info()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 35 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   50000 non-null  int64  
 1   gender               50000 non-null  str    
 2   age                  50000 non-null  int64  
 3   height_cm            50000 non-null  int64  
 4   weight_kg            50000 non-null  int64  
 5   waist_cm             50000 non-null  float64
 6   eyesight_left        50000 non-null  float64
 7   eyesight_right       50000 non-null  float64
 8   hearing_left         50000 non-null  float64
 9   hearing_right        50000 non-null  float64
 10  systolic             50000 non-null  float64
 11  relaxation           50000 non-null  float64
 12  fasting_blood_sugar  50000 non-null  float64
 13  cholesterol          50000 non-null  float64
 14  triglyceride         50000 non-null  float64
 15  hdl                  50000 non-null  float64
 1

## Estadisticos descriptivos

In [5]:
df_train.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
id,50000.0,NaN,NaN,NaN,26284.99458,16015.070567,0.0,12499.75,24999.5,40034.25,55690.0
gender,50000,2,M,31784,NaN,NaN,NaN,NaN,NaN,NaN,NaN
age,50000.0,NaN,NaN,NaN,44.1893,12.090086,20.0,40.0,40.0,55.0,85.0
height_cm,50000.0,NaN,NaN,NaN,164.652,9.208778,130.0,160.0,165.0,170.0,190.0
weight_kg,50000.0,NaN,NaN,NaN,65.8787,12.831355,30.0,55.0,65.0,75.0,135.0
waist_cm,50000.0,NaN,NaN,NaN,3.414852,0.386765,2.125,3.166667,3.416667,3.666667,5.375
eyesight_left,50000.0,NaN,NaN,NaN,0.028446,0.024007,0.000694,0.005556,0.041667,0.043056,0.38125
eyesight_right,50000.0,NaN,NaN,NaN,0.028304,0.024216,0.000694,0.005556,0.041667,0.043056,0.38125
hearing_left,50000.0,NaN,NaN,NaN,0.042736,0.006588,0.041667,0.041667,0.041667,0.041667,0.083333
hearing_right,50000.0,NaN,NaN,NaN,0.042758,0.006655,0.041667,0.041667,0.041667,0.041667,0.083333


## Valores faltantes

Verificamos si hay nulos en alguna columna.

In [6]:
nulos = df_train.isna().sum()
print('Nulos en train:')
print(nulos[nulos > 0] if nulos.sum() else 'Sin nulos')

nulos_p = df_pred.isna().sum()
print('\nNulos en pred:')
print(nulos_p[nulos_p > 0] if nulos_p.sum() else 'Sin nulos')

Nulos en train:
Sin nulos

Nulos en pred:
Sin nulos


## Duplicados

Revisamos duplicados completos y tambien ignorando el id, para ver si hay personas repetidas con distinto identificador.

In [7]:
print('Duplicados completos:', df_train.duplicated().sum())
print('Duplicados sin id   :', df_train.drop(columns=[ID_COL]).duplicated().sum())

Duplicados completos: 0


Duplicados sin id   : 5448


## Cardinalidad y columnas constantes

Buscamos columnas con un solo valor (no sirven para predecir nada) y columnas muy concentradas en un valor.

In [8]:
card = pd.DataFrame({
    'n_unique': df_train.nunique(),
    'top': df_train.apply(lambda s: s.value_counts(normalize=True).index[0]),
    'top_freq': df_train.apply(lambda s: s.value_counts(normalize=True).iloc[0].round(4)),
})
card.sort_values('n_unique')

,n_unique,top,top_freq
oral,1,Y,1.0000
gender,2,M,0.6357
hearing_right,2,0.041667,0.9738
hearing_left,2,0.041667,0.9743
dental_caries,2,0,0.7869
tartar,2,Y,0.5558
smoking,2,0,0.6334
urine_protein,6,0.041667,0.9445
height_cm,13,170,0.2047
age,14,40,0.2721


In [9]:
# Columnas constantes (1 solo valor) -> 'oral' deberia aparecer aca.
constantes = [c for c in df_train.columns if df_train[c].nunique() == 1]
print('Columnas constantes:', constantes)

# Casi-constantes (>95% en un valor).
casi = [c for c in df_train.columns
        if df_train[c].value_counts(normalize=True).iloc[0] > 0.95]
print('Casi-constantes (>95%):', casi)

Columnas constantes: ['oral']


Casi-constantes (>95%): ['hearing_left', 'hearing_right', 'oral']


## Balance del target

Miramos cuantos fumadores y no fumadores hay porque eso afecta la metrica que vamos a usar.

In [10]:
print(df_train[TARGET].value_counts())
print()
print(df_train[TARGET].value_counts(normalize=True).round(4))

smoking
0    31671
1    18329
Name: count, dtype: int64

smoking
0    0.6334
1    0.3666
Name: proportion, dtype: float64


## Guardado

Guardamos los datos ya con el formato correcto para no tener que releer el Excel en cada notebook.

In [11]:
DATA_PROCESSED.mkdir(parents=True, exist_ok=True)
df_train.to_csv(DATA_PROCESSED / 'train_clean.csv', index=False)
df_pred.to_csv(DATA_PROCESSED / 'pred_clean.csv', index=False)
print('Guardados train_clean.csv y pred_clean.csv en', DATA_PROCESSED)

Guardados train_clean.csv y pred_clean.csv en /home/mvieira/Documentos/diplomatura/entregas/tp2/data/processed


## Resumen

- Train: 50.000 filas, pred: 5.692 filas.
- `oral` vale siempre "Y" -> la sacamos antes de modelar.
- `hearing_left`, `hearing_right` y `urine_protein` casi no varian -> poco aporte esperado.
- El target esta algo desbalanceado (63% no fuma, 37% fuma), lo tenemos que tener en cuenta al entrenar.
- No hay nulos.